### DISCUSSION TITLE
Why does the creation of the BusinessClass root allow creation with parameters, while in child classes, creation is only allowed without parameters?

### DISCUSSION THREAD

**Discussion by ecaldentey** [MarimerLLC/csla#3659] [Questions]
Hello,

Both of the following methods work for a root class, but if the class is a child, only the first method works:

```c#
[Create]
private void Create([Inject] IMyDal dal)
{
    LoadProperty(MyIdProperty, dal.FetchId());
    BusinessRules.CheckRules();
}

**//This method is not reached in the case of a child business class.**
[Create] This method is not reached in the case of a child business class.
private void Create(int idSelector, [Inject] IMyDal dal)
{
    LoadProperty(MyIdProperty, dal.FetchId(idSelector));
    BusinessRules.CheckRules();
}
```

**I invoke the methods from the client with the following methods:**

```c#
public async Task<MyBusinessClass> NewMyBusinessClassAsync()
{
    return await _dataPortal.CreateAsync();
}

public async Task<MyBusinessClass> NewMyBusinessClassAsync(int idSelector)
{
    return await _dataPortal.CreateAsync(idSelector);
}
```

Is there any reason for this difference, or am I doing something wrong? 

Any help will be appreciated. 

Thank you very much.

**Comment by rockfordlhotka** [comment]
There are differences between the root and child data portal. One of these differences is that the root data portal invokes data portal operations with attributes like `Create`, while the child data portal invokes operations with attributes like `CreateChild`.

**Comment by ecaldentey** [comment]
Oops, my mistake.

In this case, what is the correct way to create and retrieve a child object in Csla 7?"

I have a child class that takes its ID from a database sequence, and for this purpose, the child class has a Create method defined as follows:

[CreateChild]
private void Create(int idSelector, [Inject] IChildDal dal)
{
	LoadProperty(ChildIdProperty, dal.FetchId(idSelector));
	BusinessRules.CheckRules();
}

To invoke this method, I understand that I should do it from a method of the parent, on the client side of the parent class, like Parent.AddChild, which calls a DataPortal method of the parent. Within that method, I invoke the following code:

var myChild = dalChild.CreateChild(idSelector);
this.ChildList.Add(myChild);

I'm not sure if I'm on the right track, and I'm not sure if it's the simplest way to have a new child object with its ID in the UI to operate with it, add its properties, and then manually add the child object to the parent's list of children. In fact, that child can, in turn, have its own children, and my idea is to create the child objects without having to retrieve the last object added to the parent's children list with code like:

var myChildCreated = Parent.ChildList.Last();

What is the correct way to create and retrieve child objects?

Any help will be appreciated.

Regards,

Esteban

**Reply by Chicagoan2016** [reply]
@ecaldentey Although the Using Csla book series was written for version 4.xx but IMO they are still relevant.
In the 'Using Csla 4, Data Access' book, the discussion of various ways to create child objects starts at p-115.

**Comment by ecaldentey** [comment]
Hello @Chicagoan2016 ,

Thank you for the suggestion. In fact, I have read the CSLA 4 book, but although in conceptual terms, the creation of the business class follows the same definition, the implementation is very different. I am addressing the issue by using the strategy from the ProjectTracker example, where a helper class 'ProjectResourceEditCreator' is created.

So, to retrieve the IDs I need for the child classes, each class is associated with a helper class Creator.

**Reply by Chicagoan2016** [reply]
@ecaldentey , our database(s) uses Identity field in SQL Server, and our child objects need the ID from the parent object (table). However, we don't have exactly 1-1 mapping between objects and tables, as recommended by Rocky).
Almost all of our Csla projects are in 4.xx (long story) but after work, I could share some (pseudo) code.

**Reply by ecaldentey** [reply]
@Chicagoan2016 In my case, it's similar. I use Oracle and Postgres sequences that I must retrieve when creating objects, and my classes are also not 1-1 with the tables. In fact, I use many views that correspond to a class, and a view can have many tables. From these tables, many have their key associated with a sequence, and therefore, a class may need to retrieve several sequences to create an object. An example of this could be a view used to create users, which, in addition to creating the user, creates a record in the people table if it doesn't exist. Not all people have an account in the system, and therefore, they don't have a user, resulting in different IDs.

**Reply by ecaldentey** [reply]
@Chicagoan2016 I use stored procedures to perform all operations in the database.

**Reply by Chicagoan2016** [reply]
@ecaldentey  I got super busy at work. Are you working with Web apps? because in stateless web apps, collection (and its children) are slightly different than in desktop apps.

**Reply by ecaldentey** [reply]
@Chicagoan2016 

Indeed, I am migrating a web application with CSLA 4.5 to 7.0.3, and I already have a working Blazor 7.0 WebAssembly hosted application with an access layer for Oracle, PostgreSQL, and SQL Server (I programmed the code generation with a generator). It has the ability to read tables and views. For views, I parse the SQL to determine the involved tables and generate classes for tables and views with root, child, readonly, command, editlist, readonlylist, and namevaluepairlist objects, as well as transaction handling. It's already functioning and will serve as the basis for the migration.

I use a single stored procedure per class, which, in addition to parameters for fields, receives one defining the action to be performed: -1 to retrieve an ID from a sequence, 0 to perform a select, 1 to perform an insert, 2 for an update, 3 for a delete, and 4 for a custom action. Another parameter, 'option', allows for multiple options within each action, meaning different selects or different updates, etc.

Lastly, the root classes handle opening and closing transactions, and the interfaces to the DAL of the classes are injected using dependency injection."

If you're interested in viewing the project's code, I can upload it to GitHub.
